// @subsquid/evm-abi/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.keccak256 = exports.AbiEvent = exports.event = exports.AbiFunction = exports.viewFun = exports.fun = exports.indexed = exports.ContractBase = void 0;
var contract_base_1 = require("./contract-base");
Object.defineProperty(exports, "ContractBase", { enumerable: true, get: function () { return contract_base_1.ContractBase; } });
var indexed_1 = require("./indexed");
Object.defineProperty(exports, "indexed", { enumerable: true, get: function () { return indexed_1.indexed; } });
var function_1 = require("./abi-components/function");
Object.defineProperty(exports, "fun", { enumerable: true, get: function () { return function_1.fun; } });
Object.defineProperty(exports, "viewFun", { enumerable: true, get: function () { return function_1.viewFun; } });
Object.defineProperty(exports, "AbiFunction", { enumerable: true, get: function () { return function_1.AbiFunction; } });
var event_1 = require("./abi-components/event");
Object.defineProperty(exports, "event", { enumerable: true, get: function () { return event_1.event; } });
Object.defineProperty(exports, "AbiEvent", { enumerable: true, get: function () { return event_1.AbiEvent; } });
__exportStar(require("./errors"), exports);
const keccak256_1 = __importDefault(require("keccak256"));
exports.keccak256 = keccak256_1.default;
